File: src/LeetCode/Algorithms/Java/Backspace_String_Compare.java
package LeetCode.Algorithms.Java;

// https://leetcode.com/problems/backspace-string-compare/
public class Backspace_String_Compare {
	public static void main(String[] args) {
		String S = "xywrrmp"; //"a#c"; //"ab#c";
		String T = "xywrrmu#p";//"b";//"ad#c";
		
		System.out.println(backspaceCompare(S, T));
	}
	
	public static boolean backspaceCompare(String S, String T) {
		return (checkString(S).equals(checkString(T)) ? true : false);
	}
	
	public static String checkString(String str) {
		StringBuffer sb = new StringBuffer();
		
		for (Character c : str.toCharArray()) {
			if (c != '#') {
				sb.append(c);
			} else if (sb.length() > 0) {
				sb.deleteCharAt(sb.length() - 1);
			}
		}
		
		return sb.toString();
	}
}

File: src/LeetCode/Algorithms/JavaScript/Backspace_String_Compare.js
/**
 * https://leetcode.com/problems/backspace-string-compare/
 * @param {string} S
 * @param {string} T
 * @return {boolean}
 */
var backspaceCompare = function(S, T) {
	/**
	 * @param {string} str
	 */
	const compareString = (str) => {
		const list = [];
		
		for (s of str.split('')) {
			(s === '#') ? list.pop() : list.push(s);
		}
		
		return list.join('');
	};
	
	return compareString(S) === compareString(T);
};

let S = "ab#c";
let T = "ad#c";

console.log(backspaceCompare(S, T));